### PR TITLE
hotfix(message/validation): optimize signer state memory usage

### PR DESCRIPTION
**Pull request by nkryuchkov**
### Changes:

- [don't allocate SeenSigners if not needed](https://github.com/ssvlabs/ssv/commit/c8c925cd3d054d3b214fca1958f823107e644180)
- [use SignersBitMask as SeenSigners key](https://github.com/ssvlabs/ssv/commit/9f703dd50658166a9cf9fbb3f8874684086d3db2)
- [store full data hash instead of full data](https://github.com/ssvlabs/ssv/commit/55c75255bc8728869155ce6b93b74ed86a809491)
- [compress MessageCounts from 48 bytes to 1](https://github.com/ssvlabs/ssv/commit/8fbb44745264760217b61516f470d55c7fb7d7aa)

### Calculations of signer state for 50K validators for `MessageCounts`:

We store the state for each validator for each of the 6 roles. For each role, we store the state, of each signer, up to 13. For each signer, we store the state of the 64 slots. Then for each slot, we store the signer state. So overall we have up to 50000 * 6 * 13 * 64 ~= 249_600_000 signer states. I'll count the maximal value but the actual value should be lower because the average amount of operators is less than 13.

So if `MessageCounts` is reduced from 48 bytes to 1 then in such case the max theoretical memory consumption should reduce from ~12 GB to ~250 MB

### Changes in Pyroscope: 

image

### CODE REVIEW

File: message/validation/utils_test.go [hunk 1]
```diff
@@ -147,75 +146,68 @@ func TestValidateConsensusMessage(t *testing.T) {
...
 			name: "RoundChangeMessage_ExceedsLimit_ReturnsError",
 			input: input{
-				counts:           &MessageCounts{RoundChange: 2},
+				seenMsgTypes:     &SeenMsgTypes{v: 0b10000},
 				signedSSVMessage: &spectypes.SignedSSVMessage{},
 				msg:              &specqbft.Message{MsgType: specqbft.RoundChangeMsgType},
-				limits:           MessageCounts{RoundChange: 0},
 			},
-			expectedError: fmt.Errorf("message is duplicated, got round change, having pre-consensus: 0, proposal: 0, prepare: 0, commit: 0, round change: 2, post-consensus: 0"),
+			expectedError: fmt.Errorf("message is duplicated, got round change, having pre-consensus: 0, proposal: 0, prepare: 0, commit: 0, round change: 1, post-consensus: 0"),
```
Comment (iurii-ssv) [inline message/validation/utils_test.go:180]: This new error message makes it a bit harder to understand what's going on because it sort of claims we've seen 1 round change in that particular test-case (not 2),

can we maybe improve this a bit ? Eg. to print `true/false` if keeping track of exact number of times is memory-heavy

File: message/validation/common_checks.go [hunk 2]
```diff
@@ -140,3 +140,16 @@ func (mv *messageValidator) validateBeaconDuty(
...
+	// Probably converting committee to a map would be faster, but since committee size is <13, this should be okay.
+	// Although, we need to change it if it appears in pprof output.
+	for i, oid := range committee {
+		if oid == signer {
+			return i
+		}
+	}
+
+	mv.logger.Panic(fmt.Sprintf("signer %v must be in committee %v", signer, committee))
+	panic("unreachable") // fix compilation issue
```
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply message/validation/common_checks.go]: @y0sher 

This must be checked in advance by message validation, so triggering `mv.logger.Panic` would mean a bug in message validation, so I think this must be left as a panic. 

Regarding the second panic, we usually prefer `mv.logger.Panic` over `panic` but the go compiler doesn't know that `mv.logger.Panic` always panics and doesn't allow it to compile. However, now I think that just `panic` should be enough because it's unreachable

**Review by iurii-ssv — APPROVED** [review]
Looks good!

File: message/validation/common_checks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iurii-ssv) [reply message/validation/common_checks.go]: I'd also probably go with `error` (not `panic`) in this case, 

I don't think it's OK to `panic` here because here we have many "actors" and it takes only 1 of them to deny service to all the rest (even though panic here would mean we have a bug, we probably don't want to let "bad" actors punish "good" actors if we do - for example, after bad code-merge we accidentally remove 1 validation code line).

In other words, it's generally bad to violate the following principle: "independent components/services/handlers/actors should function independently (meaning failure in 1 of them shouldn't affect others)".

[the remark this answers is not in the document]
Comment (nkryuchkov) [reply message/validation/common_checks.go]: @iurii-ssv 

> I don't think it's OK to panic here because here we have many "actors" and it takes only 1 of them to deny service to all the rest (even though panic here would mean we have a bug, we probably don't want to let "bad" actors punish "good" actors if we do - for example, after bad code-merge we accidentally remove 1 validation code line).

IMO accidental removal of 1 line is a critical bug, which may lead to much worse consequences than the ability to DDoS. I think we need to be able to spot this bug ASAP. If we change it to error, no one would notice it

File: message/validation/validation.go [hunk 3]
```diff
@@ -259,31 +249,23 @@ func (mv *messageValidator) getCommitteeAndValidatorIndices(msgID spectypes.Mess
...
-		operatorIDs: operators,
-		indices:     []phase0.ValidatorIndex{validator.BeaconMetadata.Index},
-		committeeID: validator.CommitteeID(),
-	}, nil
+	indices := []phase0.ValidatorIndex{validator.BeaconMetadata.Index}
+	return newCommitteeInfo(validator.CommitteeID(), operators, indices), nil
 }
 
-func (mv *messageValidator) consensusState(messageID spectypes.MessageID) *consensusState {
+func (mv *messageValidator) consensusState(messageID spectypes.MessageID, committee []spectypes.OperatorID) *consensusState {
```
Comment (MatheusFranco99) [inline message/validation/validation.go:256]: Perhaps I lost a bit the track, but how are we differentiating the consensusState for a certain PeerID?
Is this being separated on a higher level than here?
I ask this because different peers can't share the same state

Comment (nkryuchkov) [reply message/validation/validation.go:256]: @MatheusFranco99 we check peer ID just to check if it's our message, we don't use it for anything other than that

File: message/validation/common_checks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iurii-ssv) [reply message/validation/common_checks.go]: > IMO accidental removal of 1 line is a critical bug, which may lead to much worse consequences than the ability to DDoS. I think we need to be able to spot this bug ASAP. If we change it to error, no one would notice it

Hm, yes that's a trade-off, 

but it seems to me that 
- any "legitimate" message that triggers this error/panic (which can happen if we have that missing-line bug) should/will be spotted by unit/spec tests we have - I guess we can run them and check
- and any "illegitimate" message ... we don't really care about immediately spotting these

**Review by MatheusFranco99 — APPROVED** [review]
Great work!!
The issue I raised doesn't regard this PR, so approving :)

File: message/validation/common_checks.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (iurii-ssv) [reply message/validation/common_checks.go]: > Also, there are a lot of other assumptions based on previous checks in message validation, and if we add all of them, it will inflate the code a lot.

That's a good point, indeed we can't cover them all - so the question is whether we want to try and cover those that are "easily visible" (like the one we are discussing here).

> But it's impossible for any message to trigger this, it's against message validation rules. 

I understand that,

> Moreover, now all errors are covered with tests, and I want to keep this enforced, however, it's impossible to write a test for this case if it's an error because it's unreachable

Right, what I meant is to 1) comment out the validation we have (the one that prevents us from reaching this `panic`) and 2) run tests to see if they discover this panic (if they do, I think it's "safe/fine" to replace it with error).

That's a lot of effort though, I'm not sure myself it's worth it. Perhaps @y0sher has anything to add on this.

[the remark this answers is not in the document]
Comment (nkryuchkov) [reply message/validation/common_checks.go]: Now this part looks like this:
```go
func (ci *CommitteeInfo) signerIndex(signer spectypes.OperatorID) int {
	idx, ok := ci.signerIndices[signer]
	if !ok {
		panic(fmt.Sprintf("BUG: message validation must have checked that signer %v is in committee %v", signer, ci.committee))
	}

	return idx
}
```
And I tend to change it to the below (or even remove this method) to keep it consistent with other message validation code (we don't usually check invariants) because I don't like adding an error

```go
func (ci *CommitteeInfo) signerIndex(signer spectypes.OperatorID) int {
	return ci.signerIndices[signer]
}
```

File: message/validation/consensus_state.go [hunk 4]
```diff
@@ -76,15 +69,15 @@ func (os *OperatorState) Set(slot phase0.Slot, epoch phase0.Epoch, state *Signer
 }
 
 func (os *OperatorState) MaxSlot() phase0.Slot {
-	os.mu.RLock()
-	defer os.mu.RUnlock()
+	os.mu.Lock()
```
Comment (oleg-ssvlabs) [inline message/validation/consensus_state.go:72]: Was there a specific reason for replacing the RWMutex with a write-only Mutex?

Comment (nkryuchkov) [reply message/validation/consensus_state.go:72]: @oleg-ssvlabs `RWMutex` consumes more memory, and the `OperatorState` memory consumption seems to be a bottleneck in the exporter. I'd use `RWMutex` here only if we benchmark it and see a significant improvement

Comment (nkryuchkov) [reply message/validation/consensus_state.go:72]: It's 64 vs 80 bytes for each `OperatorState` IIRC, not much of a difference, but I just wanted to squeeze everything out of the state structure because we allocate a lot of them: for each validator for each role for each operator. So I guess on mainnet the total difference would be a few tens of megabytes, which is not very much.

I agree that `RWMutex` would reduce mutex block time but I think the difference wouldn't be very big. But generally it looks to me as a trade-off and since we're currently fighting with exporter memory issues, I tend to prefer to reduce the memory use

Comment (nkryuchkov) [reply message/validation/consensus_state.go:72]: @oleg-ssvlabs actually, perhaps we could remove this mutex as @moshe-blox suggested in https://github.com/ssvlabs/ssv/pull/2034#discussion_r1952617946. We have validation lock by msg ID, and the message validation doesn't have concurrent checks, so we shouldn't have any data race in `OperatorState` and `ValidatorState`

**Review by y0sher — APPROVED** [review]

File: message/validation/committee_info.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	return CommitteeInfo{
+		committeeID:      committeeID,
+		committee:        operators,
+		signerIndices:    signerIndices,
+		validatorIndices: validatorIndices,
+	}
+}
+
+// keeping the method for readability and the comment
+func (ci *CommitteeInfo) signerIndex(signer spectypes.OperatorID) int {
```
[the remark this answers is not in the document]
Comment (nkryuchkov) [reply message/validation/committee_info.go:34]: There's a comment on the line below:

```go
return ci.signerIndices[signer] // existence must be checked by ErrSignerNotInCommittee
```
